ingeniamotion/pdo.py

```diff
@@ -194,17 +194,43 @@ def __init__(
                 )
             self._notify_send_process_data = notify_send_process_data
             self._notify_receive_process_data = notify_receive_process_data
+            self._notify_exceptions = notify_exceptions
 
         def run(self) -> None:
             """Start the PDO exchange"""
             self._net.start_pdos()
+            iteration_duration = -1
             while not self._pd_thread_stop_event.is_set():
+                time_start = time.time()
                 if self._notify_send_process_data is not None:
                     self._notify_send_process_data()
-                self._net.send_receive_processdata()
+                try:
+                    self._net.send_receive_processdata()
+                except ILWrongWorkingCount as e:
+                    self._pd_thread_stop_event.set()
+                    self._net.stop_pdos()
+                    if iteration_duration == -1:
+                        iteration_duration = time.time() - time_start
+                    duration_error = ""
+                    if iteration_duration > self._refresh_rate:
+                        duration_error = (
+                            f"Last iteration took {iteration_duration * 1000:0.1f} ms which is higher"
+                            f" than the refresh rate ({self._refresh_rate * 1000:0.1f} ms). Please"
+                            " optimize the callbacks and/or increase the refresh rate."
+                        )
+                    if self._notify_exceptions is not None:
+                        e = IMException(
+                            "Stopping the PDO thread due to the following exception:"
+                            f" {e} {duration_error}"
+                        )
+                        self._notify_exceptions(e)
+                    break
                 if self._notify_receive_process_data is not None:
                     self._notify_receive_process_data()
-                time.sleep(self._refresh_rate)
+                iteration_duration = time.time() - time_start
```

Comment:
This is a little confusing. 
I would do:
```
  remaining_loop_time = self._refresh_rate - (time.time() - time_start)
  if remaining_loop_time > 0:
      time.sleep(remaining_loop_time)
```

Reply:
Done!